### PR TITLE
Microhard 1.5.testing

### PR DESCRIPTION
**Pull request by NicoK**
try to remove lock contention and therefore reduce the overhead of credit-based flow control

@pnowojski can you have a look that I did not break any assumption by removing some locks (I'm pretty sure it should be ok, though)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 1]
```diff
@@ -237,6 +237,8 @@ void notifySubpartitionConsumed() {
 	@Override
 	void releaseAllResources() throws IOException {
 		if (isReleased.compareAndSet(false, true)) {
+			// stop advertising unannounced credit (best-effort)
+			getAndResetUnannouncedCredit();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: I would drop this, since it seems like it would be hard to test and it would rise a complexity of the code, while I don't expect it to do any difference even in scenarios with many short lived jobs.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 2]
```diff
@@ -493,18 +500,20 @@ void onSenderBacklog(int backlog) throws IOException {
...
-			// race condition when onSenderBacklog and releaseAllResources running in parallel.
+			// Similar to notifyBufferAvailable(), make sure that we never add a buffer
+			// after releaseAllResources() released all buffers (see above for details).
 			if (isReleased.get() || isBlocked()) {
 				return;
 			}
 
 			numRequiredBuffers = backlog + initialCredit;
-			while (bufferQueue.getAvailableBufferSize() < numRequiredBuffers && !isWaitingForFloatingBuffers) {
+			int numAvailableBuffers = bufferQueue.getAvailableBufferSize();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: Please also revert this change. It can not have any performance effect, while it inlines some of the `bufferQueue` logic here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClient.java [hunk 3]
```diff
@@ -171,10 +171,7 @@ public void operationComplete(ChannelFuture future) throws Exception {
 	}
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClient.java:172]: Hmmm all of the changes in `[FLINK-XXXXX][network] reduce locking around credit notification` seems shaky. Removing the code that's commented "do sth to avoid deadlock" seems adventurous :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClient.java [hunk 4]
```diff
@@ -171,10 +171,7 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
 
 	public void notifyCreditAvailable(RemoteInputChannel inputChannel) {
-		// We should skip the notification if the client is already closed.
-		if (!closeReferenceCounter.isDisposed()) {
-			clientHandler.notifyCreditAvailable(inputChannel);
-		}
+		clientHandler.notifyCreditAvailable(inputChannel);
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClient.java:174]: It seems to me like this `if` check was not protecting us from any race condition, there was no other synchronisation between `#notifyCreditAvailable` and `#close()` - `notifyCreditAvailable` could execute this check first and freeze, then `close()` is called, disposes everything and exits and after that `notifyCreditAvailable` could resume working.

On the other hand, this `if` check was saving us some work only in rare case of disposing channels in an exchange of synchronisation cost on common path.

Am I right?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 5]
```diff
@@ -300,8 +300,7 @@ protected void setBlocked(boolean isBlocked) {
 	private void notifyCreditAvailable() {
 		checkState(partitionRequestClient != null, "Tried to send task event to producer before requesting a queue.");
 
-		// We should skip the notification if this channel is already released.
-		if (!isReleased.get() && !isBlocked()) {
+		if (!isBlocked()) {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: What happens if `notifyCreditAvailable` is carried on after releasing? Add unit test for this?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 6]
```diff
@@ -365,13 +364,6 @@ boolean isWaitingForFloatingBuffers() {
 	 */
 	@Override
 	public boolean notifyBufferAvailable(Buffer buffer) {
-		// Check the isReleased state outside synchronized block first to avoid
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:356]: I do not understand what `deadlock`/situation was this comment mentioning and that makes me uncomfortable with removing it. Do you understand what was it for?

**Review by pnowojski — CHANGES_REQUESTED** [review]
Besides comments left in code I do not like the general idea of merging/shipping on a short notice commit: `
[FLINK-XXXXX][network] reduce locking around credit notification`

Those changes would require quite long "burning" time on travis to catch all possible race conditions

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java (hunk 6 above)
Comment (NicoK) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:356]: I also did not see anything that can lead to a deadlock (hence the removal) - also, this condition could be true right after the if-block, so this is also not really protecting against anything (if a deadlock ever existed)

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java (hunk 5 above)
Comment (NicoK) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: Because here, we could also release right after the check or any time before the handling of this call in the netty thread, the netty thread needs to handle this case anyway.

A new unit test for this specific scenario seems reasonable though.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java (hunk 1 above)
Comment (NicoK) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: I would actually like to go one step further, because I don't like this part of the code in `CreditBasedPartitionRequestClientHandler` which is called for each credit announcement:
```
int unannouncedCredit = inputChannel.getAndResetUnannouncedCredit();
if (unannouncedCredit == 0 || inputChannel.isReleased()) {
```
-> I don't want it to access `AtomicInteger unannouncedCredit` and `AtomicBoolean isReleased` and it would be best it only one would suffice - or maybe this is (unnecessary) micro-optimisation?
I'll check whether my benchmarks would react on such a change...
